stdpopsim/slim_engine.py

```diff
@@ -0,0 +1,663 @@
+        :type slim_script: bool
+        :param slim_scaling_factor: Rescale model parameters by the given value,
+            to speed up simulation. Population sizes and generation times are
+            divided by this factor, whereas the mutation rate, recombination
+            rate, and growth rates are multiplied by the factor.
+            See SLiM manual: `5.5 Rescaling population sizes to improve
+            simulation performance.`
+        :type slim_scaling_factor: float
+        :param slim_no_recapitation: Do an explicit burn in, and add
+            mutations, within the SLiM simulation. This may be much slower than
+            the defaults (recapitation and neutral mutation overlay with
+            msprime). The burn in behaviour is to wait until all individuals in
+            the ancestral populations have a common ancestor within their
+            respective population, and then wait another 10*N generations.
+        :type slim_no_recapitation: bool
+        :param slim_no_burnin: Do not perform a burn in at the start of the
+            simulation.  This option is only relevant when
+            ``slim_no_recapitation=True``.
+        :type slim_no_burnin: bool
+        """
+
+        run_slim = not slim_script
+        do_recap = not slim_no_recapitation
+        check_coalescence = slim_no_recapitation and not slim_no_burnin
+
+        if slim_path is None:
+            slim_path = self.slim_path()
+
+        if do_recap:
+            mutation_rate = contig.mutation_rate
+            # Ensure no mutations are introduced by SLiM.
+            contig = stdpopsim.Contig(
+                    recombination_map=contig.recombination_map,
+                    mutation_rate=0,
+                    genetic_map=contig.genetic_map)
+
+        slim_cmd = [slim_path]
+        if seed is not None:
+            slim_cmd.extend(["-s", f"{seed}"])
+
+        mktemp = functools.partial(tempfile.NamedTemporaryFile, mode="w")
+
+        @contextlib.contextmanager
+        def script_file_f():
+            f = mktemp(suffix=".slim") if not slim_script else sys.stdout
+            yield f
+            # Don't close sys.stdout.
+            if not slim_script:
+                f.close()
+
+        with script_file_f() as script_file, mktemp(suffix=".ts") as ts_file:
+
+            slim_makescript(
+                    script_file, ts_file.name,
+                    demographic_model, contig, samples,
+                    slim_scaling_factor, check_coalescence, verbosity)
+
+            script_file.flush()
+
+            if not run_slim:
+                return None
+
+            slim_cmd.append(script_file.name)
+            stdout = subprocess.DEVNULL if verbosity == 0 else None
+            subprocess.check_call(slim_cmd, stdout=stdout)
+
+            ts = pyslim.load(ts_file.name)
+
+        if do_recap:
+            rng = random.Random(seed)
+            s1, s2 = rng.randrange(1, 2**32), rng.randrange(1, 2**32)
```

Comment:
we are going to want to be able to set these seeds externally

Reply:
I'm not sure exactly what you mean here. The seeds are inherited from the `seed` parameter passed to `engine.simulate()`, so simulations should be deterministic when that parameter is used.

Reply:
Yes, this is fine, resolving.